codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/aws/AwsJson10ProtocolGenerator.java

```diff
@@ -86,10 +97,156 @@ private void generateSharedDeserializers(GenerationContext context, GoWriter wri
+                            "awsQueryCompatible", context.getService().hasTrait(AwsQueryCompatibleTrait.class)
+                                    ? deserializeAwsQueryError()
+                                    : emptyGoTemplate(),
+                            "errors", GoWriter.ChainWritable.of(
+                                    operation.getErrors(context.getService()).stream()
+                                            .map(it ->
+                                                    deserializeErrorCase(context, context.getModel().expectShape(
+                                                            it, StructureShape.class)))
+                                            .toList()
+                            ).compose(false)
+                    )));
+        }
     }
 
+    private GoWriter.Writable deserializeErrorCase(GenerationContext ctx, StructureShape error) {
+        return goTemplate("""
+                case $type:S:
+                    verr, err := $deserialize:L(v)
+                    if err != nil {
+                        return &$deserError:T{
+                            Err: $fmtErrorf:T("deserialize $type:L: %w", err),
+                            Snapshot: payload,
+                        }
+                    }
+                    $awsQueryCompatible:W
+                    return verr
+                """,
+                MapUtils.of(
+                        "deserError", SmithyGoDependency.SMITHY.pointableSymbol("DeserializationError"),
+                        "deserialize", getDeserializerName(error),
+                        "equalFold", SmithyGoDependency.STRINGS.func("EqualFold"),
+                        "fmtErrorf", GoStdlibTypes.Fmt.Errorf,
+                        "type", error.getId().toString(),
+                        "awsQueryCompatible", ctx.getService().hasTrait(AwsQueryCompatibleTrait.class)
+                                ? deserializeModeledAwsQueryError()
+                                : emptyGoTemplate()
+                ));
+    }
+
+    private GoWriter.Writable deserializeAwsQueryError() {
+        return goTemplate("""
+                if qtype := getAwsQueryErrorCode(resp); len(qt) > 0 {
+                    typ = qtype
+                }""");
+    }
+
+    private GoWriter.Writable deserializeModeledAwsQueryError() {
+        return goTemplate("""
+                if qtype := getAwsQueryErrorCode(resp); len(qt) > 0 {
+                    verr.ErrorCodeOverride = $T(qtype)
+                }""", SmithyGoTypes.Ptr.String);
+    }
+
+    private GoWriter.Writable getProtocolErrorInfo() {
+        return goTemplate("""
+            func getProtocolErrorInfo(payload []byte) (typ, msg string, v $value:T, err error) {
+
+                paid := $reader:T(payload)
+                jsonDecoder := $decoder:T(paid)
+                var val interface{}
+                var jv map[string]interface{}
+
+                jsonDecoder.Decode(&val)
```

Comment:
This also returns an error that we should check for

Reply:
I thought that any errors from the "decode" call would have been caught by the jv map error check, and that it didn't matter the type that the decoder would be unloading into, but I suppose that there could also be an error that arises from this that I'm not thinking of. I'll go ahead and add it anyways though.